node/lib/util/stash_util.js

```diff
@@ -310,30 +443,84 @@ exports.apply = co.wrap(function *(repo, id, reinstateIndex) {
+        let subRepo =
             yield opener.getSubrepo(name,
                                     Open.SUB_OPEN_OPTION.FORCE_OPEN);
 
-        // Try to get the comit for the stash; if it's missing, fail.
-
+        // Try to get the commit for the stash; if it's missing, fail.
+        let stashCommit;
         try {
-            yield subRepo.getCommit(stashSha);
-        }
-        catch (e) {
+            stashCommit = yield Commit.lookup(subRepo, stashSha);
+        } catch (e) {
             console.error(`\
 Stash commit ${colors.red(stashSha)} is missing from submodule \
 ${colors.red(name)}`);
             result = null;
             return;                                                   // RETURN
         }
 
+        const stashTree = yield stashCommit.getTree();
+
+        const worktreeTree = yield stashCommit.getTree();
+        const indexCommit = yield stashCommit.parent(1);
+        const indexTree = yield indexCommit.getTree();
+
+        if (stashCommit.parentcount() > 2) {
+            const oldHead = yield stashCommit.parent(2);
+            if (stashCommit.parentcount() > 3) {
+                const stagedCommit = yield stashCommit.parent(3);
+                index[name] = stagedCommit.id();
+            }
+
+            // Before we get started, we might need to rebase the
+            // commits from oldHead..commitBeforeStash
+
+            const commitBeforeStash = yield stashCommit.parent(0);
+
+            const rebaseError = function(resolution) {
+                console.error(`The stash for submodule ${name} had one or \
+more commits, ending with ${commitBeforeStash.id()}.  We tried to rebase these \
+commits onto the current commit, but this failed.  ${resolution}
+After you are done with this rebase, you may need to apply working tree \
+and index changes.  To restore the index, try (inside ${name}) 'git read-tree \
+${indexCommit.id()}'.  To restore the working tree, try (inside ${name}) \
+'git checkout ${stashCommit} -- .'`);
+            };
+
+            try {
+                const res = yield SubmoduleRebaseUtil.rewriteCommits(
+                    subRepo,
+                    commitBeforeStash,
+                    oldHead);
+                if (res.errorMessage) {
+                    rebaseError(res.errorMessage);
+                    result = null;
+                    return;
+                }
+            } catch (e) {
+                // We expect these errors to be caught, but if
+                // something goes wrong, wWe are leaving the user in a
+                // pretty yucky state.  the alternative is to try to
+                // back out the whole stash apply, which seems worse.
+
+                rebaseError(`We are leaving the rebase half-finished, so \
+that you can fix the conflicts and continue by running 'git rebase \
+--continue' inside of ${name} (note: not 'git meta rebase').`);
+                console.error(`The underlying error, which is likely useless \
```

Comment:
I think user usually is not interested in knowing about nodegit, would you make the message more user friendly, maybe something like "unkown error but unlikely harmful"?

Reply:
(changed)